File: javascript/finished/scripts/setup-store.ts
/**
 * Setup script: Create FileSearchStore and upload MLH documentation
 * 
 * This script:
 * 1. Creates a FileSearchStore in Google AI
 * 2. Downloads MLH docs from GitHub
 * 3. Uploads them to the store
 * 4. Polls until indexing is complete
 * 
 * Run: npm run setup-store
 */

import { GoogleGenAI } from '@google/generative-ai';
import { STORE_NAME, MLH_DOCS } from '../lib/store';
import * as fs from 'fs';
import * as path from 'path';

const API_KEY = process.env.GOOGLE_API_KEY;

if (!API_KEY) {
  console.error('❌ GOOGLE_API_KEY environment variable is required');
  console.error('Get your key from https://aistudio.google.com/apikey');
  process.exit(1);
}

const genai = new GoogleGenAI(API_KEY);

/**
 * Download a file from URL to a local temp path
 */
async function downloadFile(url: string, filename: string): Promise<string> {
  console.log(`  📥 Downloading ${filename}...`);
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to download ${url}: ${response.statusText}`);
  }
  
  const content = await response.text();
  const tmpDir = path.join(process.cwd(), 'tmp');
  
  // Create tmp directory if it doesn't exist
  if (!fs.existsSync(tmpDir)) {
    fs.mkdirSync(tmpDir, { recursive: true });
  }
  
  const filepath = path.join(tmpDir, filename);
  fs.writeFileSync(filepath, content, 'utf-8');
  console.log(`  ✅ Saved to ${filepath}`);
  return filepath;
}

/**
 * Create a FileSearchStore
 */
async function createStore(name: string) {
  console.log(`\n🏗️  Creating FileSearchStore: ${name}`);
  
  try {
    // Check if store already exists
    const existingStore = await genai.fileSearchStores.get(name);
    console.log(`✅ Store "${name}" already exists`);
    return existingStore;
  } catch (error) {
    // Store doesn't exist, create it
    console.log(`  Creating new store...`);
    const store = await genai.fileSearchStores.create({
      displayName: name,
    });
    console.log(`✅ Store created: ${store.name}`);
    return store;
  }
}

/**
 * Upload files to the store
 */
async function uploadFiles(storeName: string, filepaths: string[]) {
  console.log(`\n📤 Uploading ${filepaths.length} files...`);
  
  const uploadedFiles = [];
  
  for (const filepath of filepaths) {
    const filename = path.basename(filepath);
    console.log(`  Uploading ${filename}...`);
    
    const fileContent = fs.readFileSync(filepath);
    
    // Upload file to Google AI
    const file = await genai.files.upload({
      file: {
        data: fileContent,
        mimeType: 'text/markdown',
      },
      displayName: filename,
    });
    
    console.log(`  ✅ Uploaded: ${file.name}`);
    
    // Add file to store
    await genai.fileSearchStores.addFile(storeName, file.name);
    console.log(`  ➕ Added to store`);
    
    uploadedFiles.push(file.name);
  }
  
  return uploadedFiles;
}

/**
 * Poll store state until indexing is complete
 */
async function waitForIndexing(storeName: string) {
  console.log(`\n⏳ Waiting for indexing to complete...`);
  
  let attempts = 0;
  const maxAttempts = 60; // 5 minutes max
  
  while (attempts < maxAttempts) {
    const store = await genai.fileSearchStores.get(storeName);
    
    if (store.state === 'ACTIVE') {
      console.log(`✅ Indexing complete! Store is ACTIVE`);
      return;
    }
    
    console.log(`  State: ${store.state} (attempt ${attempts + 1}/${maxAttempts})`);
    
    // Wait 5 seconds before next check
    await new Promise(resolve => setTimeout(resolve, 5000));
    attempts++;
  }
  
  throw new Error('Indexing timed out after 5 minutes');
}

/**
 * Test the store with a sample query
 */
async function testQuery(storeName: string) {
  console.log(`\n🧪 Testing store with a sample query...`);
  
  const model = genai.getGenerativeModel({
    model: 'gemini-2.0-flash',
    tools: [
      {
        fileSearchTool: {
          fileSearchStore: storeName,
        },
      },
    ],
  });
  
  const result = await model.generateContent(
    'What are the key responsibilities of an MLH hackathon organizer?'
  );
  
  const response = result.response;
  console.log(`\n💬 Query: What are the key responsibilities of an MLH hackathon organizer?`);
  console.log(`\n📝 Response:\n${response.text()}`);
  
  // Show citations if available
  if (response.candidates?.[0]?.groundingMetadata?.groundingChunks) {
    const chunks = response.candidates[0].groundingMetadata.groundingChunks;
    console.log(`\n📚 Citations: ${chunks.length} source(s)`);
  }
}

/**
 * Main setup flow
 */
async function main() {
  console.log('🚀 MLH RAG Store Setup\n');
  console.log(`Store name: ${STORE_NAME}`);
  console.log(`Documents: ${MLH_DOCS.length}\n`);
  
  try {
    // Step 1: Create store
    const store = await createStore(STORE_NAME);
    
    // Step 2: Download docs
    console.log(`\n📥 Downloading MLH documentation...`);
    const filepaths: string[] = [];
    
    for (const doc of MLH_DOCS) {
      const filepath = await downloadFile(doc.url, `${doc.name}.md`);
      filepaths.push(filepath);
    }
    
    // Step 3: Upload to store
    await uploadFiles(store.name, filepaths);
    
    // Step 4: Wait for indexing
    await waitForIndexing(store.name);
    
    // Step 5: Test with a query
    await testQuery(store.name);
    
    // Cleanup temp files
    console.log(`\n🧹 Cleaning up temp files...`);
    for (const filepath of filepaths) {
      fs.unlinkSync(filepath);
    }
    fs.rmdirSync(path.join(process.cwd(), 'tmp'));
    
    console.log(`\n✨ Setup complete! Your store is ready to use.`);
    console.log(`\nNext steps:`);
    console.log(`  1. Copy .env.example to .env`);
    console.log(`  2. Add your GOOGLE_API_KEY`);
    console.log(`  3. Run: npm run dev`);
    
  } catch (error) {
    console.error(`\n❌ Setup failed:`, error);
    process.exit(1);
  }
}

main();

File: python/finished/setup_store.py
#!/usr/bin/env python3
"""
Setup script to create a Gemini FileSearchStore and upload MLH documentation.

This script:
1. Creates a new FileSearchStore
2. Downloads MLH docs from GitHub (raw markdown files)
3. Uploads them to the store
4. Polls until indexing is complete
5. Prints the store name for use in your .env file

Run this once before using the chat app:
    python setup_store.py
"""

import os
import time
import tempfile
import requests
from pathlib import Path
from dotenv import load_dotenv
from google import genai
from google.genai import types

# Load environment variables
load_dotenv()

# MLH documentation URLs (raw GitHub content)
# Note: mlh-hackathon-organizer-guide uses 'master' branch, others use 'main'
MLH_DOCS = {
    "hackathon-organizer-guide.md": "https://raw.githubusercontent.com/MLH/mlh-hackathon-organizer-guide/master/README.md",
    "what-is-mlh.md": "https://raw.githubusercontent.com/MLH/mlh-hackathon-organizer-guide/master/overview/what-is-mlh.md",
    "hackathon-timeline.md": "https://raw.githubusercontent.com/MLH/mlh-hackathon-organizer-guide/master/general-information/hackathon-timeline.md",
    "getting-sponsorship.md": "https://raw.githubusercontent.com/MLH/mlh-hackathon-organizer-guide/master/general-information/getting-sponsorship/README.md",
    "managing-registrations.md": "https://raw.githubusercontent.com/MLH/mlh-hackathon-organizer-guide/master/general-information/managing-registrations/README.md",
    "judging-and-submissions.md": "https://raw.githubusercontent.com/MLH/mlh-hackathon-organizer-guide/master/general-information/judging-and-submissions/README.md",
    "mlh-community-values.md": "https://raw.githubusercontent.com/MLH/mlh-hackathon-organizer-guide/master/overview/mlh-community-values.md",
    "code-of-conduct.md": "https://raw.githubusercontent.com/MLH/mlh-policies/main/code-of-conduct.md",
    "community-values.md": "https://raw.githubusercontent.com/MLH/mlh-policies/main/community-values.md",
    "hack-days-guide.md": "https://raw.githubusercontent.com/MLH/mlh-hack-days-organizer-guide/main/README.md",
}


def download_file(url: str, filename: str, temp_dir: Path) -> Path:
    """Download a file from URL to temp directory."""
    print(f"  📥 Downloading {filename}...")
    response = requests.get(url)
    response.raise_for_status()

    file_path = temp_dir / filename
    file_path.write_text(response.text, encoding="utf-8")
    print(f"     ({len(response.text)} chars)")
    return file_path


def wait_for_indexing(client: genai.Client, store_name: str, timeout: int = 300):
    """Poll the store until all documents are indexed."""
    print("\n⏳ Waiting for indexing to complete...")
    start_time = time.time()

    while time.time() - start_time < timeout:
        store = client.file_search_stores.get(name=store_name)

        active = store.active_documents_count or 0
        pending = store.pending_documents_count or 0
        failed = store.failed_documents_count or 0
        total = active + pending + failed

        print(f"  Progress: {active} active, {pending} pending, {failed} failed (total: {total})")

        if pending == 0 and total > 0:
            if failed > 0:
                print(f"  ⚠️  {failed} documents failed to index")
            print("✅ Indexing complete!")
            return True

        time.sleep(5)

    print("⚠️  Timeout waiting for indexing")
    return False


def main():
    # Initialize Gemini client
    api_key = os.getenv("GEMINI_API_KEY")
    if not api_key:
        print("❌ Error: GEMINI_API_KEY not found in .env file")
        print("   Get your key from https://aistudio.google.com/apikey")
        return

    client = genai.Client(api_key=api_key)

    # Create FileSearchStore
    print("📦 Creating FileSearchStore...")
    store = client.file_search_stores.create(
        config=types.CreateFileSearchStoreConfig(
            display_name="MLH Documentation Store",
        )
    )
    print(f"✅ Created store: {store.name}")

    # Download and upload files
    print("\n📄 Downloading and uploading MLH docs...\n")
    with tempfile.TemporaryDirectory() as temp_dir:
        temp_path = Path(temp_dir)

        for filename, url in MLH_DOCS.items():
            try:
                # Download file
                file_path = download_file(url, filename, temp_path)

                # Upload directly to the FileSearchStore
                print(f"  📤 Uploading {filename} to store...")
                operation = client.file_search_stores.upload_to_file_search_store(
                    file_search_store_name=store.name,
                    file=str(file_path),
                )

                # Wait for upload operation to complete
                while not operation.done:
                    time.sleep(2)
                    operation = client.operations.get(operation)

                if operation.error:
                    print(f"  ❌ Error uploading {filename}: {operation.error}")
                else:
                    print(f"  ✅ {filename} uploaded successfully\n")

            except Exception as e:
                print(f"  ❌ Error processing {filename}: {e}\n")

    # Wait for indexing
    wait_for_indexing(client, store.name)

    # Print final instructions
    print("\n" + "=" * 60)
    print("🎉 Setup complete!")
    print("=" * 60)
    print(f"\nStore Name: {store.name}")
    print("\n📝 Next steps:")
    print("1. Copy the store name above")
    print("2. Add it to your .env file:")
    print(f"   FILE_SEARCH_STORE_NAME={store.name}")
    print("3. Test from the command line:")
    print(f'   python query_test.py "How do I get reimbursed for a Hack Day?"')
    print("4. Or run the chat app:")
    print("   streamlit run app.py")
    print()


if __name__ == "__main__":
    main()
